---
title: "Estimation of scaling factor and equilibrium constants"
output: html_notebook
---
```{r}
## Chunk for knitting outside of Github
library(ezknitr)
ezknit(file = "eps_and_Keq_estimation_final.Rmd", out_dir = "~/AvOwork")
```

```{r}
setwd("~/mnt/")
source("~/HexamerBias/deltaGprediction/binding_model_functions.r")
source('~/HexamerBias/rscripts/hexamer_sequence_functions.r')
```

<!-- Loading required files: tab of Mfold binding energies, tab of groups of simmetric hexamers, genomic kmer abundances for 3 genomes, primer-template tables -->
<!-- ```{r} -->
<!-- tabDg <- read.delim(gzfile("~/mnt/edann/hexamers/rand_hex_deltaG_ions.txt.gz"), sep=' ', header = FALSE, col.names=c('template', 'dG')) -->
<!-- t.groups <- read.table("~/AvOwork/MatchingTemplate_groups.tsv", col.names = c('template', 'group')) -->

<!-- kmer.ab.cele <- load.kmer.abundance('./edann/hexamers/genomes_kmers/WBcel235.kmerAbundance.csv') -->
<!-- kmer.ab.zf <- load.kmer.abundance('./edann/hexamers/genomes_kmers/danRer10.kmerAbundance.csv') -->
<!-- kmer.ab.human <- load.kmer.abundance('./edann/hexamers/genomes_kmers/hg38.kmerAbundance.csv') -->

<!-- cele.all <- load.pt.data("./edann/VAN2423_onePreamp/cov_prediction/CG-pbat-gDNA-CeleTotal-noBS-1preAmp-handMix_lmerged_R1.ptCounts.qualFilt.csv") -->
<!-- cele.all.df <- join.pt.data(cele.all$matches, cele.all$t.usage, kmer.ab.cele, tabDg) -->

<!-- zf.all <- load.pt.data("./edann/VAN2423_onePreamp/cov_prediction/CG-pbat-gDNA-zfishTotal-noBS-1preAmp-handMix_lmerged_R1.ptCounts.qualFilt.csv") -->
<!-- zf.all.df <- join.pt.data(zf.all$matches, zf.all$t.usage, kmer.ab.zf, tabDg) -->

<!-- human.all <- load.pt.data("./edann/VAN2423_onePreamp/cov_prediction/CG-pbat-gDNA-humanAPKS-noBS-1preAmp-handMix_lmerged_R1.ptCounts.qualFilt.csv") -->
<!-- human.all.df <- join.pt.data(human.all$matches, human.all$t.usage, kmer.ab.human, tabDg) -->
<!-- ``` -->

<!-- Saving RData files for future use -->
<!-- ```{r} -->
<!-- save(human.all.df, file = "~/AvOwork/human_pt_all.RData") -->
<!-- save(zf.all.df, file = "~/AvOwork/zfish_pt_all.RData") -->
<!-- ``` -->

```{r}
load("~/AvOwork/rdata/human_pt_all.RData")
load("~/AvOwork/rdata/cele_pt_all.RData")
load("~/AvOwork/rdata/zfish_pt_all.RData")
```

<!-- ## Before epsilon -->
<!-- In the previous version of the model, I predicted binding free energies for the primer-template binding reaction as follows: -->
<!-- $$ -->
<!-- exp(\Delta G_{ij})=\frac{p_it_j}{\sum_j{\sum_j{p_it_j}}}\cdot\frac{\sum_j{T_j}}{T_j - \sum_i{p_it_j}} -->
<!-- $$ -->
<!-- Here the normalization strategy is inappropriate, as $p_it_j$ count and the free template count ($T_j - \sum_i{p_it_j}$) are normalized on different factors.    -->

<!-- Recently we observed we could compute reasonable $\Delta$G values also removing the normalization factors and using the differnt variables as simple counts: -->
<!-- $$ -->
<!-- exp(\Delta G_{ij})=\frac{p_it_j}{T_j - \sum_i{p_it_j}} -->
<!-- $$ -->
<!-- ```{r} -->
<!-- cele.df %>% -->
<!--   filter(pt>200) %>% -->
<!--   mutate(pred.deltaG=pt/(abundance-t.usage)) %>% -->
<!--   ggplot(., aes(dG, log(pred.deltaG))) + -->
<!--   geom_point(alpha=0.3) -->


<!-- ``` -->



<!-- ## Estimation of scaling factor epsilon -->
<!-- We expect that the primer-template pairs that we measure are proportional but not equal to the real amount of pairs $[p_it_j]$ that we would expect by binding primers on one copy of the genome. That is because each sample might have a different copy number for the genome (because of a different number of cells, or loss of a chromosome etc...) and different sequencing depth. For this reason we define: -->
<!-- $$ -->
<!--   x_{ij}=\varepsilon[p_it_j], -->
<!-- $$ -->
<!-- where $\varepsilon$ is a scaling-factor that accounts for all sample specific effects. -->


<!-- #### Chi-square on multiple groups  -->
<!-- Using the chi-square approach on groups of sequences with the same $\Delta$G I estimate epsilon sampling 20 groups with at least 20 sequences from every genome.  -->

<!-- I can estimate epsilon using as primer concentration the probability distribution from a simulated random pool (requires taking the average of more iterations), or assuming the same concentration for each primer ($4/4^6$). As I don't see a significant difference between the two approaches, I chose to proceed assuming even primer concentration, which makes the estimation quicker. -->

<!-- <!-- ```{r} --> -->
<!-- <!-- cele.eps <-epsilon.iterative(cele.df, estimation.it = 40) --> -->
<!-- <!-- cele.eps.prob <-epsilon.iterative(cele.df, estimation.it = 30, sample.size = 10, imposeP = T) --> -->

<!-- <!-- # zf.eps <-epsilon.iterative(zf.df, estimation.it = 20) --> -->
<!-- <!-- # zf.eps.prob <-epsilon.iterative(zf.df, estimation.it = 40, sample.size = 10, imposeP = T) --> -->

<!-- <!-- mutate(melt(cele.eps, id.vars='n', variable.name='iteration', value.name = 'eps'), species='cele.dist') %>% --> -->
<!-- <!--   bind_rows(., mutate(melt(cele.eps.prob, id.vars='n', variable.name='iteration', value.name = 'eps'), species='cele.prob')) %>% --> -->
<!-- <!--   # bind_rows(., mutate(melt(zf.eps, id.vars='n', variable.name='iteration', value.name = 'eps'), species='zf')) %>% --> -->
<!-- <!--   ggplot(., aes(eps, group=species, fill=species)) +  --> -->
<!-- <!--   facet_grid(species~.) + --> -->
<!-- <!--   geom_histogram()   --> -->

<!-- <!-- # mutate(melt(zf.eps, id.vars='n', variable.name='iteration', value.name = 'eps'), species='zf.dist') %>% --> -->
<!-- <!-- #   bind_rows(., mutate(melt(zf.eps.prob, id.vars='n', variable.name='iteration', value.name = 'eps'), species='zf.prob')) %>% --> -->
<!-- <!-- #   # bind_rows(., mutate(melt(zf.eps, id.vars='n', variable.name='iteration', value.name = 'eps'), species='zf')) %>% --> -->
<!-- <!-- #   ggplot(., aes(eps, group=species, fill=species)) +  --> -->
<!-- <!-- #   facet_grid(species~.) + --> -->
<!-- <!-- #   geom_histogram()   --> -->
<!-- <!-- ``` --> -->

<!-- ##### Estimation for all species with fixed probability -->
<!-- ```{r} -->
<!-- cele.eps <-epsilon.iterative(select.diag.pairs(cele.all.df), tot.it = 300, estimation.it = 1, sample.size = 10, imposeP = T) -->
<!-- zf.eps <-epsilon.iterative(select.diag.pairs(zf.all.df), tot.it = 300, estimation.it = 1, sample.size = 10, imposeP = T) -->
<!-- human.eps <-epsilon.iterative(select.diag.pairs(human.all.df), tot.it = 300,estimation.it = 1, sample.size = 10, imposeP = T) -->
<!-- ``` -->

<!-- ```{r, echo=F} -->
<!-- mutate(melt(zf.eps, id.vars='n', variable.name='iteration', value.name = 'eps'), species='zf') %>% -->
<!--   bind_rows(., mutate(melt(cele.eps, id.vars='n', variable.name='iteration', value.name = 'eps'), species='cele')) %>% -->
<!--   bind_rows(., mutate(melt(human.eps, id.vars='n', variable.name='iteration', value.name = 'eps'), species='human')) %>% -->
<!--   ggplot(., aes(eps, group=species, fill=species)) +  -->
<!--   facet_grid(species~.) + -->
<!--   xlab('epsilon') + -->
<!--   theme(axis.title=element_text(size=20), -->
<!--         axis.text=element_text(size=16), -->
<!--         strip.text=element_text(size=20), -->
<!--         legend.text = element_text(size=20), -->
<!--         legend.title = element_text(size=22)) + -->
<!--   geom_histogram()   -->

<!-- # ggsave("~/AvOwork/output/epsilon_estimate/eps_dist_groupChi_fixedP.pdf") -->
<!-- ``` -->

<!-- Interestingly the size of the sample of groups I take for the estimation affects the range of epsilons -->
<!-- ```{r} -->
<!-- cele.eps.3 <-epsilon.iterative(select.diag.pairs(cele.all.df), estimation.it = 20, sample.size = 3,imposeP = T) -->
<!-- cele.eps.4 <-epsilon.iterative(select.diag.pairs(cele.all.df), estimation.it = 20, sample.size = 4,imposeP = T) -->
<!-- cele.eps.5 <-epsilon.iterative(select.diag.pairs(cele.all.df), estimation.it = 20, sample.size = 5,imposeP = T) -->
<!-- cele.eps.10 <-epsilon.iterative(select.diag.pairs(cele.all.df), estimation.it = 20, sample.size = 10,imposeP = T) -->
<!-- cele.eps.15 <-epsilon.iterative(select.diag.pairs(cele.all.df), estimation.it = 20, sample.size = 15,imposeP = T) -->
<!-- cele.eps.20 <-epsilon.iterative(select.diag.pairs(cele.all.df), estimation.it = 20, sample.size = 20,imposeP = T) -->
<!-- cele.eps.25 <-epsilon.iterative(select.diag.pairs(cele.all.df), estimation.it = 20, sample.size = 25,imposeP = T) -->
<!-- cele.eps.30 <-epsilon.iterative(select.diag.pairs(cele.all.df), estimation.it = 20, sample.size = 30,imposeP = T) -->

<!-- mutate(melt(cele.eps.5, id.vars='n', variable.name='iteration', value.name = 'eps'), sample.size=5) %>% -->
<!--   # bind_rows(., mutate(melt(cele.eps.3, id.vars='n', variable.name='iteration', value.name = 'eps'), sample.size=3)) %>% -->
<!--   # bind_rows(., mutate(melt(cele.eps.4, id.vars='n', variable.name='iteration', value.name = 'eps'), sample.size=4)) %>% -->
<!--   bind_rows(., mutate(melt(cele.eps.10, id.vars='n', variable.name='iteration', value.name = 'eps'), sample.size=10)) %>% -->
<!--   bind_rows(., mutate(melt(cele.eps.15, id.vars='n', variable.name='iteration', value.name = 'eps'), sample.size=15)) %>% -->
<!--   bind_rows(., mutate(melt(cele.eps.20, id.vars='n', variable.name='iteration', value.name = 'eps'), sample.size=20)) %>% -->
<!--   bind_rows(., mutate(melt(cele.eps.25, id.vars='n', variable.name='iteration', value.name = 'eps'), sample.size=25)) %>% -->
<!--   bind_rows(., mutate(melt(cele.eps.30, id.vars='n', variable.name='iteration', value.name = 'eps'), sample.size=30)) %>% -->
<!-- # bind_rows(., mutate(melt(zf.eps, id.vars='n', variable.name='iteration', value.name = 'eps'), species='zf')) %>% -->
<!--   ggplot(., aes(sample.size, eps, group=sample.size, fill=sample.size)) +  -->
<!--   # facet_grid(species~.) + -->
<!--   geom_violin() + -->
<!--   NULL -->
<!-- ``` -->

<!-- This observation made me try using the Chi-square approach on one group at a time to estimate epsilon.  -->

<!-- #### Chi-square on single groups  -->
<!-- The distribution is less defined (also because there are no iterations involved) -->
<!-- ```{r} -->
<!-- epsilons.cele <- sapply(seq(1,40), function(x) estimate.eps.one.group(cele.df, group = x)) -->
<!-- epsilons.human <- sapply(seq(1,40), function(x) estimate.eps.one.group(human.df, group = x)) -->
<!-- epsilons.zf <- sapply(seq(1,40), function(x) estimate.eps.one.group(zf.df, group = x)) -->

<!-- df.eps.single <- mutate(data.frame(t(epsilons.cele)), eps= as.numeric(eps), group=as.numeric(group)) %>% -->
<!--   inner_join(., mutate(data.frame(t(epsilons.human)), eps= as.numeric(eps), group=as.numeric(group)), by='group', suffix=c('.cele', '.human')) %>% -->
<!--   inner_join(., mutate(data.frame(t(epsilons.zf)), eps= as.numeric(eps), group=as.numeric(group)), by='group') %>% -->
<!--   rename(eps.zf=eps)  -->

<!-- df.eps.single %>% -->
<!--   melt(id.vars='group', variable.name='species', value.name='eps') %>% -->
<!--   ggplot(., aes(eps, fill=species)) + -->
<!--   facet_grid(species~.) + -->
<!--   geom_histogram(bins=40) -->
<!-- ``` -->

<!-- Also, the group of hexamers on which the estimation is made influences the outcome consistently between species -->
<!-- ```{r} -->
<!-- df.eps.single %>% -->
<!--   ggplot(., aes(eps.cele, eps.human, label=group)) + geom_text() + -->
<!--   theme(axis.title = element_text(size=20)) -->
<!-- ``` -->

<!-- ### Correlating epsilon with % of genome covered -->
<!-- Whatever the method of choice, the value of epsilon obtained is correlated with the overall genomic coverage. -->
<!-- I use ```bedtools genomecov``` to compute the percentage of genome covered for each experiment. -->

<!-- ```{r} -->
<!-- cov <- data.frame(species = c('cele', 'human', 'zf'), cov=c(1-0.0638419,1-0.754045,1-0.603038)) -->

<!-- species.epsilons<- mutate(melt(zf.eps, id.vars='n', variable.name='iteration', value.name = 'eps'), species='zf') %>% -->
<!--   bind_rows(., mutate(melt(cele.eps, id.vars='n', variable.name='iteration', value.name = 'eps'), species='cele')) %>% -->
<!--   bind_rows(., mutate(melt(human.eps, id.vars='n', variable.name='iteration', value.name = 'eps'), species='human')) %>%  -->
<!--   # melt(id.vars='group', variable.name='species', value.name='eps') %>% -->
<!--   mutate(species=gsub(species, pattern = 'eps.', replacement = '')) %>% -->
<!--   group_by(species) %>% -->
<!--   summarise(eps=median(eps, na.rm=T)) -->

<!-- species.eps.onegroup <- mutate(data.frame(groups=unlist(epsilons.zf[1,]), eps=unlist(epsilons.zf[2,])) , species='zf') %>% -->
<!--   bind_rows(., mutate(data.frame(groups=unlist(epsilons.human[1,]), eps=unlist(epsilons.human[2,])) , species='human')) %>% -->
<!--   bind_rows(., mutate(data.frame(groups=unlist(epsilons.cele[1,]), eps=unlist(epsilons.cele[2,])) , species='cele')) -->

<!-- species.eps.onegroup %>% -->
<!--   inner_join(., cov, by='species') %>% -->
<!--   filter(groups<10) %>% -->
<!--   ggplot(., aes(cov, eps, group=groups, color=groups)) + -->
<!--   # facet_wrap(~groups, scales='free_y') + -->
<!--   geom_point() + geom_line() -->

<!-- species.epsilons %>% inner_join(., cov, by='species') %>% -->
<!--   ggplot(., aes(cov,eps)) + geom_point() + -->
<!--   geom_smooth(method='loess') -->
<!-- ``` -->

<!-- (When the cluster is back) We could use a downsampling approach to see if computing epsilon on less reads we still get the same regression line, to be used to estimate epsilon in different experiments. -->


<!-- ## Estimation of $K_{eq}$ -->

<!-- Using the estimated value for epsilon, I compute the assciation constant $K_{eq}$ for template $j$ and primer $i$ as: -->
<!-- $$ -->
<!--   K_{ij} = [p_i]\left(\frac{\varepsilon T_j - \sum_ix_{ij}}{x_{ij}}\right) -->
<!-- $$ -->
<!-- Where $[p_i] = \frac{4}{4^6}$, $T_j$ is the fraction of genomic abundance for the template, $x_{ij}$ is the count of occurrencies for the primer-template pair $ij$. To improve the statistical power of the estimation, I assume that the association constant for primer $i$ on template $j$ is the same as the association constant for primer $j$ on template $i$, thereby taking the average of these two values. -->

<!-- First I filter out low count primer-template pairs (non relevant binding events) -->
<!-- ```{r} -->
<!-- cele.filt.df <- filter(cele.all.df, pt > 300) -->
<!-- human.filt.df <- filter(human.all.df, pt > 300) -->
<!-- zf.filt.df <- filter(zf.all.df, pt > 300) -->
<!-- ``` -->

<!-- Then I estimate $K_{eq}$ for all pairs -->
<!-- ```{r} -->
<!-- cele.keqs <- compute.keqs.fixedP(cele.filt.df, mean.eps = median(as.matrix(cele.eps[,-1]), na.rm=T), take.pairs = T) -->
<!-- zf.keqs <- compute.keqs.fixedP(zf.filt.df, mean.eps = median(as.matrix(zf.eps[,-1]), na.rm=T), take.pairs = T) -->
<!-- human.keqs <- compute.keqs.fixedP(human.filt.df, mean.eps = median(as.matrix(human.eps[,-1]), na.rm=T), take.pairs = T) -->
<!-- ``` -->

<!-- Comparison between $K_{eq}$ in different species -->
<!-- ```{r} -->
<!-- keqs <- zf.keqs %>% -->
<!--   dplyr::select(template, primer, pair,single.keq) %>% -->
<!--   inner_join(., dplyr::select(cele.keqs, template, primer, pair, single.keq), by=c('template', 'primer'), suffix=c('.zf', '.cele')) %>% -->
<!--   inner_join(., dplyr::select(human.keqs, template, primer, pair, single.keq), by=c('template', 'primer')) %>% -->
<!--   rename(single.keq.human=single.keq) %>% -->
<!--   group_by(pair) %>% -->
<!--   summarise(single.keq.human=mean(single.keq.human),single.keq.cele=mean(single.keq.cele), single.keq.zf=mean(single.keq.zf)) %>% -->
<!--   ungroup() %>% -->
<!--   mutate(nuc=sapply(pair, prevalent_nucleotide)) %>% -->
<!--   rename(K.human=single.keq.human, -->
<!--          K.cele=single.keq.cele, -->
<!--          K.zfish=single.keq.zf) -->

<!-- ``` -->

<!-- ```{r, echo=F} -->
<!-- cors <- cor(dplyr::select(keqs, K.human, K.cele, K.zfish) ) -->

<!-- keqs %>%   -->
<!--   ggplot(., aes(log(-K.cele), log(-K.zfish))) + -->
<!--   geom_point(alpha=0.4, aes(color=nuc)) + -->
<!--   theme_classic() + -->
<!--   geom_abline(slope=1, intercept=0, color='red') + -->
<!--   annotate('text',-4,-7,label=paste("PCC = ",round(cors['K.zfish','K.cele'],3)), size=6) + -->
<!--   theme(axis.title=element_text(size=20), -->
<!--         legend.position = 'bottom', -->
<!--         legend.title = element_text(size=20), -->
<!--         legend.text = element_text(size=18)) + -->
<!--   scale_color_discrete(name="Prevalent nucleotide") + -->
<!--   NULL -->

<!-- keqs %>%   -->
<!--   ggplot(., aes(log(-K.human), log(-K.zfish))) + -->
<!--   geom_point(alpha=0.4, aes(color=nuc)) + -->
<!--   theme_classic() + -->
<!--   geom_abline(slope=1, intercept=0, color='red') + -->
<!--   annotate('text',-5,-7,label=paste("PCC = ",round(cors['K.zfish','K.human'],3)), size=6) + -->
<!--     theme(axis.title=element_text(size=20), -->
<!--         legend.position = 'bottom', -->
<!--         legend.title = element_text(size=20), -->
<!--         legend.text = element_text(size=18)) + -->
<!--   scale_color_discrete(name="Prevalent nucleotide") + -->
<!--   NULL -->

<!-- keqs %>%   -->
<!--   ggplot(., aes(log(-K.cele), log(-K.human))) + -->
<!--   geom_point(alpha=0.4, aes(color=nuc)) + -->
<!--   theme_classic() + -->
<!--   geom_abline(slope=1, intercept=0, color='red') + -->
<!--   annotate('text',-5,-7,label=paste("PCC = ",round(cors['K.human','K.cele'],3)), size=6) + -->
<!--     theme(axis.title=element_text(size=20), -->
<!--         legend.position = 'bottom', -->
<!--         legend.title = element_text(size=20), -->
<!--         legend.text = element_text(size=18)) + -->
<!--   scale_color_discrete(name="Prevalent nucleotide") + -->
<!--   NULL -->
<!-- ``` -->

<!-- Taking as final values mean between species -->
<!-- ```{r} -->
<!-- keqs.mean <- keqs %>% -->
<!--   dplyr::select(-nuc) %>% -->
<!--   # dplyr::select(pair,single.keq.cele, single.keq.human, single.keq.zf) %>% -->
<!--   melt(id.vars="pair", variable.name='species', value.name='keq') %>% -->
<!--   group_by(pair) %>% -->
<!--   summarise(mean.keq=mean(keq, na.rm=T), sd.keq=sd(keq, na.rm=T)) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- write.csv(x = keqs.mean, "~/Keqs_all_pairs.csv") -->
<!-- ``` -->

<!-- #### Correlation of Keq with Mfold deltaG -->

<!-- ```{r} -->
<!-- pred.keq.cele <- add.pairs.info(cele.df) %>% -->
<!--   inner_join(., keqs.mean, by='pair')  -->
<!-- pred.keq.cele <- pred.keq.cele %>% -->
<!--     mutate(nuc=sapply(template, prevalent_nucleotide), -->
<!--          p=4/(4^6))  -->
<!-- pred.keq.cele %>% -->
<!--   mutate(nuc=sapply(template, prevalent_nucleotide), -->
<!--          p=4/(4^6)) %>% -->
<!--   ggplot(., aes(dG, log(-mean.keq/p), color=nuc)) + -->
<!--   geom_point(alpha=0.3) + -->
<!--       theme(axis.title=element_text(size=20), -->
<!--         legend.position = 'bottom', -->
<!--         legend.title = element_text(size=20), -->
<!--         legend.text = element_text(size=18)) + -->
<!--   xlab('DeltaG MFold') + -->
<!--   scale_color_discrete(name="Prevalent nucleotide") + -->
<!--   NULL -->

<!-- ``` -->

<!-- One immediate observation is that the range for the predicted $\Delta$G is quite smaller than the MFold. -->

<!-- ## Coverage prediction in no-BS -->
<!-- I use my model and the predicted $K_{eq}$ values and epsilon to estimate template usage (coverage) as: -->
<!-- $$ -->
<!-- C^\prime_j = \varepsilon \cdot T_j\frac{\sum_i\frac{[p_i]}{K_{ij}}}{1+\sum_i\frac{[p_i]}{K_{ij}}} -->
<!-- $$ -->

<!-- ```{r} -->
<!-- pred.keq.cele %>% -->
<!--   mutate(abundance=abundance/sum(abundance)) %>% -->
<!--   mutate(phi=p/-mean.keq, -->
<!--          nuc=sapply(template, prevalent_nucleotide), -->
<!--          epsilon=median(as.matrix(cele.eps[,-1]), na.rm=T)) %>% -->
<!--   group_by(template) %>% -->
<!--   summarise(abundance=first(abundance), epsilon=first(epsilon), t.usage=first(t.usage), sum.phi=sum(phi), nuc=first(nuc)) %>% -->
<!--   mutate(pred.cov=epsilon*(abundance*(sum.phi/(1+sum.phi)))) %>% -->
<!--   ggplot(., aes(log(t.usage/sum(t.usage)), log(pred.cov/sum(pred.cov)), color=nuc)) + geom_point(alpha=0.2) + -->
<!--   xlab("log(expected cov)") + ylab('log(predicted cov)') + -->
<!--     scale_color_discrete(name="Prevalent nucleotide") + -->
<!--         theme(axis.title=element_text(size=20), -->
<!--         # legend.position = 'bottom', -->
<!--         legend.title = element_text(size=20), -->
<!--         legend.text = element_text(size=18)) + -->
<!--   NULL -->

<!-- ``` -->

<!-- ## Coverage prediction in BS-seq experiments -->

<!-- I estimate epsilon for the BS experiment based on the % of genomic coverage in the experiment.  -->
<!-- ```{r} -->
<!-- load("~/AvOwork/rdata/D3R_pt.RData") -->
<!-- lm.epsilon <- species.epsilons %>%  -->
<!--   inner_join(., cov, by='species') %>% -->
<!--   lm(eps ~ cov, data=.) -->

<!-- perc.d3R <- 1-0.951344 -->
<!-- eps.d3r <- predict.lm(lm.epsilon,data.frame(cov=c(perc.d3R)))  -->

<!-- d3r.filt.df <- d3r.all.df %>% filter(pt>10)  -->
<!-- d3r.keqs <- compute.keqs.fixedP(d3r.filt.df, mean.eps = eps.d3r, take.pairs = T) -->

<!-- ``` -->

<!-- <!-- Reproducibility of binding events between BS-experiments --> -->
<!-- <!-- ```{r} --> -->
<!-- <!-- join.DR <- d2r.all.df %>% --> -->
<!-- <!--   filter(pt!=0) %>% --> -->
<!-- <!--   dplyr::select(primer, template, t.usage, pt, abundance) %>% --> -->
<!-- <!--   inner_join(., dplyr::select(d3r.filt,primer, template, t.usage, pt), by=c('primer', 'template'), suffix=c('.D2R', '.D3R'))  --> -->
<!-- <!-- head(join.DR) --> -->
<!-- <!-- join.DR %>% --> -->
<!-- <!--   ggplot(., aes(log(t.usage.D2R), log(t.usage.D3R))) +  --> -->
<!-- <!--   geom_point(alpha=0.3) --> -->
<!-- <!--   # ggplot(., aes(t.usage.D2R, t.usage.D3R)) + geom_point(alpha=0.3) --> -->
<!-- <!-- ``` --> -->

<!-- <!-- Then I can pool the two repliate experiments --> -->
<!-- <!-- ```{r} --> -->
<!-- <!-- pooled.DR <- join.DR %>% --> -->
<!-- <!--   mutate(pt=pt.D2R+pt.D3R, t.usage=t.usage.D2R+t.usage.D3R) %>% --> -->
<!-- <!--   dplyr::select(primer, template, pt,t.usage, abundance) --> -->

<!-- <!-- pooled.DR %>% --> -->
<!-- <!--   filter(pt>30) %>% --> -->
<!-- <!--   nrow --> -->

<!-- <!-- join.DR %>% --> -->
<!-- <!--   filter(grepl("C", template))  --> -->

<!-- <!-- dr.pooled.keqs <- compute.keqs.fixedP(filter(pooled.DR, pt>10), mean.eps = eps.d3r, take.pairs = F)  --> -->

<!-- <!-- dr.pooled.keqs %>% --> -->
<!-- <!--   mutate(p = 4/4^6) %>% --> -->
<!-- <!--   # mutate(abundance=abundance/sum(abundance)) %>% --> -->
<!-- <!--   mutate(phi=p/-single.keq) %>% --> -->
<!-- <!--   # head --> -->
<!-- <!--     group_by(template) %>% --> -->
<!-- <!--   summarise(abundance=first(frac.abundance), epsilon=first(epsilon*2), t.usage=first(t.usage), sum.phi=sum(phi)) %>% --> -->
<!-- <!--   mutate(pred.cov=epsilon*(abundance*(sum.phi/(1+sum.phi)))) %>% --> -->
<!-- <!--   ggplot(., aes(log(t.usage), log(pred.cov))) + geom_point(alpha=0.2) --> -->
<!-- <!-- ``` --> -->

<!-- When predicting the coverage first of all I find templates with a negative predicted coverage. -->
<!-- ```{r} -->
<!-- d3r.keqs %>% -->
<!--   # mutate(p = 4/4^6) %>% -->
<!--   filter(pt>10) %>% -->
<!--   # mutate(abundance=abundance/sum(abundance)) %>% -->
<!--   mutate(phi=p/-single.keq) %>% -->
<!--   # ggplot(., aes(phi)) + geom_histogram(bins=40) -->
<!--     group_by(template) %>% -->
<!--   summarise(abundance=first(frac.abundance), epsilon=first(epsilon), t.usage=first(t.usage), sum.phi=sum(phi)) %>% -->
<!--     # ggplot(., aes(sum.phi)) + geom_histogram(bins=40) -->
<!--   # group_by(template) %>% -->
<!--   # # mutate(pred.cov=epsilon*(abundance*(sum.phi/(1+sum.phi)))) %>% -->
<!--   mutate(pred.cov=epsilon*(abundance*(sum.phi/(1+sum.phi)))) %>% -->
<!--   mutate(nuc=sapply(template,prevalent_nucleotide), -->
<!--          CG=ifelse(grepl(pattern = 'CG', template), 'y', 'n')) %>% -->
<!--   # ggplot(., aes(log(t.usage/sum(t.usage)), log(abundance/sum(abundance)))) +  -->
<!--   ggplot(., aes((t.usage/sum(t.usage)), (pred.cov/sum(pred.cov)))) + -->
<!--   theme_classic() + -->
<!--   geom_point(alpha=0.4, aes(color=nuc, shape=CG)) + -->
<!--     # ggplot(., aes(log(t.usage), log(pred.cov))) +  -->
<!--   xlab("expected cov") + ylab('predicted cov') + -->
<!--     scale_color_discrete(name="Prevalent nucleotide") + -->
<!--         theme(axis.title=element_text(size=20), -->
<!--         # legend.position = 'bottom', -->
<!--         legend.title = element_text(size=20), -->
<!--         legend.text = element_text(size=18)) + -->
<!--   geom_abline(slope=1, intercept=0, color='red')  -->
<!-- ```  -->

<!-- This has to do with the value of epsilon. Trying a smaller value of epsilon (absolutely not in line with the genomic coverage hypothesis) solves the issue but the correlation is still poor. -->
<!-- ```{r} -->
<!-- d3r.keqs <- compute.keqs.fixedP(d3r.filt.df, mean.eps = 1000, take.pairs = T) -->

<!-- d3r.keqs %>% -->
<!--   filter(pt>10) %>% -->
<!--   mutate(phi=p/-single.keq) %>% -->
<!--     group_by(template) %>% -->
<!--   summarise(abundance=first(frac.abundance), epsilon=first(epsilon), t.usage=first(t.usage), sum.phi=sum(phi)) %>% -->
<!--   mutate(pred.cov=epsilon*(abundance*(sum.phi/(1+sum.phi)))) %>% -->
<!--   mutate(nuc=sapply(template,prevalent_nucleotide), -->
<!--          CG=ifelse(grepl(pattern = 'CG', template), 'y', 'n')) %>% -->
<!--   ggplot(., aes(log(t.usage/sum(t.usage)), log(pred.cov/sum(pred.cov)))) + -->
<!--   theme_classic() + -->
<!--   geom_point(alpha=0.4, aes(color=nuc, shape=CG)) + -->
<!--   xlab("log(expected cov)") + ylab('log(predicted cov)') + -->
<!--     scale_color_discrete(name="Prevalent nucleotide") + -->
<!--         theme(axis.title=element_text(size=20), -->
<!--         # legend.position = 'bottom', -->
<!--         legend.title = element_text(size=20), -->
<!--         legend.text = element_text(size=18)) + -->
<!--   geom_abline(slope=1, intercept=0, color='red')  -->
<!-- ``` -->

<!-- All in all, given the dependence of the epsilon estimation on the sampling and groups and these issues in the BS-seq estimation, I would say that this approach for estimating epsilon and $K_{eq}$ is really introducing a lot of numerical errors that we could exclude using relative energies.x -->

<!-- ## MORE ON THIS -->


<!-- ```{r} -->
<!-- epsilons.cele <- sapply(seq(1,40), function(x) estimate.eps.one.group(cele.df, group = x)) -->
<!-- epsilons.human <- sapply(seq(1,40), function(x) estimate.eps.one.group(human.df, group = x)) -->
<!-- epsilons.zf <- sapply(seq(1,40), function(x) estimate.eps.one.group(zf.df, group = x)) -->

<!-- species.eps.onegroup <- mutate(data.frame(groups=unlist(epsilons.zf[1,]), eps=unlist(epsilons.zf[2,])) , species='zf') %>% -->
<!--   bind_rows(., mutate(data.frame(groups=unlist(epsilons.human[1,]), eps=unlist(epsilons.human[2,])) , species='human')) %>% -->
<!--   bind_rows(., mutate(data.frame(groups=unlist(epsilons.cele[1,]), eps=unlist(epsilons.cele[2,])) , species='cele')) -->

<!-- epsil.onegroup <- species.eps.onegroup %>% group_by(species) %>% summarise(eps=median(eps, na.rm = T)) -->


<!-- epsil.onegroup %>% -->
<!--   inner_join(., cov, by='species') %>% -->
<!-- # Ã‡  filter(groups<10) %>% -->
<!--   ggplot(., aes(cov, eps)) + -->
<!--   # facet_wrap(~groups, scales='free_y') + -->
<!--   geom_point() + geom_line() -->
<!-- ``` -->

<!-- ```{r} -->

<!-- cele.keqs <- compute.keqs.fixedP.totabundance(cele.filt.df, mean.eps = filter(epsil.onegroup, species=='cele')$eps, take.pairs = T) -->
<!-- zf.keqs <- compute.keqs.fixedP.totabundance(zf.filt.df, mean.eps = filter(epsil.onegroup, species=='zf')$eps, take.pairs = T) -->
<!-- human.keqs <- compute.keqs.fixedP.totabundance(human.filt.df, mean.eps = filter(epsil.onegroup, species=='human')$eps, take.pairs = T) -->

<!-- keqs <- zf.keqs %>% -->
<!--   dplyr::select(template, primer, pair,single.keq) %>% -->
<!--   inner_join(., dplyr::select(cele.keqs, template, primer, pair, single.keq), by=c('template', 'primer'), suffix=c('.zf', '.cele')) %>% -->
<!--   inner_join(., dplyr::select(human.keqs, template, primer, pair, single.keq), by=c('template', 'primer')) %>% -->
<!--   rename(single.keq.human=single.keq) %>% -->
<!--   group_by(pair) %>% -->
<!--   summarise(single.keq.human=mean(single.keq.human),single.keq.cele=mean(single.keq.cele), single.keq.zf=mean(single.keq.zf)) %>% -->
<!--   ungroup() %>% -->
<!--   mutate(nuc=sapply(pair, prevalent_nucleotide)) %>% -->
<!--   rename(K.human=single.keq.human, -->
<!--          K.cele=single.keq.cele, -->
<!--          K.zfish=single.keq.zf) -->
<!-- ``` -->


## YET ANOTHER EPSILON ESTIMATION
We search for the value of the scaling factor $\epsilon$ that minimizes the difference between the expected association constant ($e^{-\beta\Delta G}$, taking $\Delta$G values from calorimetry). To avoid numerical errors in the estimation we transform all factors to logarithms. 
$$
\chi^2 = \sum{_{ij} - log([p]) + log\left(T_j - \frac{\sum_i{x_{ij}}}{\epsilon}\right) + log\left(\frac{x_{ij}}{\epsilon}\right)}
$$

Where we fix that the concentration of total template should also be higher than the concentration of bound template:
$$
T_j - \frac{\sum_i{x_{ij}}}{\epsilon} > 0
\epsilon > \frac{\sum_i{x_{ij}}}{T_j}
$$

```{r}
epsilon.minimize.chisq <- function(pt.df, max, min=0, plot=T){
  min.epsilon <- pt.df %>%
    mutate(ep=t.usage/abundance) %>%
    top_n(n = 1,ep) %>%
    .$ep
  chis <- c()
  if(min < min.epsilon){min <- min.epsilon}
  for (eps in seq(min, max, 1) ) {
    chi.sq <- pt.df %>%
      filter(pt!=0) %>%
      mutate(chi=(-dG/0.59)
             - log(4/(4^6)) 
             - log(abundance-(t.usage/eps)) 
             + log(pt/eps)
             ) %>%
      summarise(chi=sum(chi^2))
    chis <- c(chis, chi.sq)  
  }
  if(plot){plot(seq(min, max,1), unlist(chis), pch='.')}
  best.eps <- seq(min, max, 1)[which.min(unlist(chis))] 
  return(best.eps)
  }
```

```{r}
human.eps <- epsilon.minimize.chisq(human.df, max=10000)
zf.eps <- epsilon.minimize.chisq(zf.df, max=15000)
cele.eps <- epsilon.minimize.chisq(cele.df, min=130000, max=140000, plot=F)
```

#### Correlation with genomic coverage
I compute percentage of genome covered using `bedtools genomecov`

```{r}
cov %>%
  bind_cols(eps=c(eps.cele, eps.human, eps.zf)) %>%
  ggplot(., aes(log(cov), log(eps), label=species)) + 
  geom_point() +
  geom_text_repel()
```

We will confirm this trend by downsampling and adding the new samples.

#### Keq estimation
```{r}
compute.keqs <- function(pt.df, eps, filter.pt=200){
  keqs <- pt.df %>%
    filter(pt>filter.pt) %>%
    mutate(keq=((4^6)/4)*(pt/(eps*abundance-t.usage))) 
  return(keqs)
  }

human.keqs <- compute.keqs(human.df, eps.human, filter.pt = 20)
zf.keqs <- compute.keqs(zf.df, eps.zf, filter.pt = 20)
cele.keqs <- compute.keqs(cele.df, eps.cele, filter.pt = 20)
```

Correlation between predicted keqs
```{r}
inner_join(zf.keqs, human.keqs, by=c('template', 'primer'), suffix=c('.zf', '.human')) %>%
  ggplot(., aes(log(keq.zf), log(keq.human))) +
  geom_point(alpha=0.3) +
  theme_classic() +
  geom_abline(slope=1, intercept=0, color='red')

inner_join(cele.keqs, human.keqs, by=c('template', 'primer'), suffix=c('.cele', '.human')) %>%
  ggplot(., aes(log(keq.cele), log(keq.human))) +
  geom_point(alpha=0.3) +
  theme_classic() +
  geom_abline(slope=1, intercept=0, color='red')

inner_join(cele.keqs, zf.keqs, by=c('template', 'primer'), suffix=c('.cele', '.zf')) %>%
  ggplot(., aes(log(keq.zf), log(keq.cele))) +
  geom_point(alpha=0.3) +
  theme_classic() +
  geom_abline(slope=1, intercept=0, color='red')

```


#### Coverage prediction
```{r}
predict.coverage <- function(keqs.df, prob=4/(4^6)){
  pred.cov <- keqs.df %>%
    mutate(p=prob) %>%
    mutate(phi=p*keq,
           nuc=sapply(template, prevalent_nucleotide),
           epsilon=eps.human) %>%
    group_by(template) %>%
    summarise(abundance=first(abundance), epsilon=first(epsilon), t.usage=first(t.usage), sum.phi=sum(phi), nuc=first(nuc)) %>%
    mutate(pred.cov=epsilon*abundance*(sum.phi/1+sum.phi)) 
  return(pred.cov)
  }

# human.keqs <- compute.keqs(human.df, eps.human, filter.pt = 50)
# zf.keqs <- compute.keqs(zf.df, eps.zf, filter.pt = 50)
# cele.keqs <- compute.keqs(cele.df, eps.cele, filter.pt = 50)

predict.coverage(human.keqs) %>%
  mutate(nuc=sapply(template, prevalent_nucleotide)) %>%
    ggplot(., aes(log(t.usage/sum(t.usage)), log(pred.cov/sum(pred.cov)), color=nuc)) + 
  geom_point(alpha=0.4) +
  geom_abline(slope=1, intercept=0, color='red') +
  theme_classic() +
  xlab("expected cov") + ylab("predicted cov") +
  ggtitle("Human - no BS") 

predict.coverage(zf.keqs) %>%
  mutate(nuc=sapply(template, prevalent_nucleotide)) %>%
    ggplot(., aes(log(t.usage/sum(t.usage)), log(pred.cov/sum(pred.cov)), color=nuc)) + 
  geom_point(alpha=0.4) +
  geom_abline(slope=1, intercept=0, color='red') +
  theme_classic() +
  xlab("expected cov") + ylab("predicted cov") +
  ggtitle("Zebrafish - no BS")

predict.coverage(cele.keqs) %>%
  mutate(nuc=sapply(template, prevalent_nucleotide)) %>%
    ggplot(., aes(log(t.usage/sum(t.usage)), log(pred.cov/sum(pred.cov)), color=nuc)) + 
  geom_point(alpha=0.4) +
  geom_abline(slope=1, intercept=0, color='red') +
  theme_classic() +
  xlab("expected cov") + ylab("predicted cov") +
  ggtitle("C.elegans - no BS")




```

There is a general underestimation of A rich hexamers. This is expected because we filter out primer-template events that occur less than a set number of times and we've seen that A rich hexamers tend to be bound weakly by a wide range of hexamers.

### Estimate epsilon in BS-seq experiments
To estimate espilon in BS-seq experiments we use the same Chi-square approach as described abouve but using only templates that do not contain any Cs, that are not affected by bisulfite conversion. 
```{r}
load("~/AvOwork/rdata/D3R_pt.RData")

d3r.noC <- d3r.all.df %>%
  filter(template==primer) %>%
  filter(!grepl(template, pattern = "C"))
  
epsilon.minimize.chisq(d3r.noC, max=1000)
```

Is this value in line with genomic coverage? I redo epsilon estimation for the non converted samples using only hexamers with no Cs.
```{r}
perc.d3R <- 1-0.951344

human.noC <- human.df %>%
   filter(!grepl(template, pattern = "C")) 

zf.noC <- zf.df %>%
   filter(!grepl(template, pattern = "C")) 

cele.noC <- cele.df %>%
   filter(!grepl(template, pattern = "C")) 

chi.human.onlyC <- epsilon.minimize.chisq(human.noC, max=10000)
chi.zf.onlyC <- epsilon.minimize.chisq(zf.noC, max=10000)
chi.cele.onlyC <- epsilon.minimize.chisq(cele.noC, max=15000)

cov %>%
  bind_rows(data.frame(species="mouse.BS", cov=perc.d3R)) %>%
  bind_cols(eps=c(chi.cele.onlyC, chi.human.onlyC, chi.zf.onlyC, 14)) %>%
  ggplot(., aes(log(cov), log(eps))) + 
  geom_point() +
  geom_text_repel(aes(label=species))

```


#### Coverage prediction in BS-seq
```{r}
d3r.keqs <- d3r.all.df %>%
  filter(pt>10) %>%
  mutate(keq=((4^6)/4)*(pt/(14*abundance-t.usage))) 

predict.coverage(d3r.keqs) %>%
  mutate(nuc=sapply(template, prevalent_nucleotide)) %>%
    ggplot(., aes(log(t.usage/sum(t.usage)), log(pred.cov/sum(pred.cov)), color=nuc)) + 
  geom_point(alpha=0.4) +
  geom_abline(slope=1, intercept=0, color='red') +
  theme_classic() +
  xlab("expected cov") + ylab("predicted cov") +
  ggtitle("Mouse - BS")


```

Probably with less reads you also have less unspecific binding.

#### Coverage prediction with variable primer concentration
```{r}

```




<!-- # Selection of filtering cut-off -->
<!-- Based on reproducibility of primer-template counts across different biological replicates -->
<!-- ```{r} -->
<!-- d2r.diag.df <- select.diag.pairs(d2r.all.df)  -->
<!-- d2r.diag.df <- select.diag.pairs(d2r.all.df)  -->
<!-- dr.diag <- d3r.all.df %>% -->
<!--   dplyr::select(template, primer, pt) %>% -->
<!--   inner_join(., dplyr::select(d2r.all.df, template, primer, pt), by=c('template', 'primer'), suffix=c(".D3", '.D2'))  -->

<!-- dr.diag %>%   -->
<!--   ggplot(., aes(log(pt.D2/sum(pt.D2)), log(pt.D3/sum(pt.D3)))) + geom_point(alpha=0.3) + -->
<!--     theme_classic() + -->
<!--     coord_fixed(ratio=1) + -->
<!--     # geom_smooth(method='lm', color='blue') + -->
<!--     geom_abline(slope=1, intercept=0, color='red') -->

<!-- dr.diag %>% -->
<!--   mutate(pt.D2= pt.D2/sum(pt.D2), pt.D3= pt.D3/sum(pt.D3)) %>% -->
<!--   filter(pt.D2!=0 & pt.D3!=0) %>% -->
<!--   mutate(a=(pt.D3-pt.D2)/pt.D2) %>% -->
<!--   melt(id.vars=c('template', 'primer', 'a'), variable.name='rep', value.name='pt') %>% -->
<!--   ggplot(., aes(pt, a)) +  -->
<!--   facet_grid(rep~.) + -->
<!--   geom_point() + -->
<!--   geom_vline(xintercept =0.0025, color='blue', linetype=5) + -->
<!--   ylab("Residuals") + xlab("pt fraction") -->

<!-- ``` -->



<!-- ## Estimation from BS seq experiments -->
<!-- Loading data -->
<!-- ```{r} -->
<!-- kmer.ab.mouse <- load.kmer.abundance('~/mnt/edann/hexamers/genomes_kmers/mm10.kmerAbundance.csv') -->

<!-- d2r.all <- load.pt.data("~/mnt/edann/primer_conc_VAN2493/D2R_tr2_R1_bismark_bt2.ptCounts.qualFilt.csv") -->
<!-- d2r.all.df <- join.pt.data(d2r.all$matches, d2r.all$t.usage, kmer.ab.mouse, tabDg) -->

<!-- d3r.all <- load.pt.data("~/mnt/edann/primer_conc_VAN2493/D3R_tr2_R1_bismark_bt2.ptCounts.qualFilt.csv") -->
<!-- d3r.all.df <- join.pt.data(d3r.all$matches, d3r.all$t.usage, kmer.ab.mouse, tabDg) -->

<!-- d3t.all <- load.pt.data("~/mnt/edann/primer_conc_VAN2493/D3T_tr2_R1_bismark_bt2.ptCounts.qualFilt.csv") -->
<!-- d3t.all.df <- join.pt.data(d3r.all$matches, d3r.all$t.usage, kmer.ab.mouse, tabDg) -->


<!-- ``` -->

<!-- I want to prove that epsilon is proportional to the genomic coverage (how much of the genome do you cover). -->

<!-- Let's see how Keq changes for different values of epsilon -->
<!-- ```{r} -->

<!-- variable.eps <- function(pt.df, eps){ -->
<!--   cele.keqs <-   mutate(pt.df,   -->
<!--                   p=4/(4^6), -->
<!--                   epsilon=eps, -->
<!--                   frac.abundance=(abundance/sum(as.numeric(abundance))) -->
<!--                   ) %>% -->
<!--     mutate(single.keq=p*((epsilon*(frac.abundance/pt))-(t.usage/pt)))  -->
<!--     # summarize(mean.k=mean(single.keq)) -->
<!--   chi <- cele.keqs %>% -->
<!--     mutate(chi=dG-log(-single.keq)) %>% -->
<!--     summarize(chi=sum(chi)) -->
<!--   return(c(e,cele.keqs$mean.k)) -->
<!--   } -->

<!-- chi.sq <- c() -->
<!-- for (e in sapply(-10:2, function(x) 10^x)) { -->
<!--   chi.sq <- rbind(chi.sq, variable.eps(cele.filt.df, e))   -->
<!-- } -->

<!-- as.data.frame(chi.sq) %>% -->
<!--   rename(eps=V1, chi.sq=V2) %>% -->
<!-- # ggplot(., aes(log10(log10.eps), log(-chi.sq+1))) + -->
<!--   ggplot(., aes(eps, -chi.sq)) + -->
<!--   geom_line() + -->
<!--   theme_classic() -->

<!-- chi.sq -->
<!-- ``` -->



<!-- Coverage prediction in no-BS -->
<!-- ```{r} -->
<!-- cele.pred.cov <- cele.keqs %>% -->
<!--   mutate(abundance=abundance/sum(abundance)) %>% -->
<!--   # mutate(phi=p/-single.keq) %>% -->
<!--   mutate(phi=p/-single.keq) %>% -->
<!--   # ggplot(., aes(phi)) + geom_histogram(bins=40) -->
<!--   group_by(template) %>% -->
<!--   summarise(abundance=first(frac.abundance), epsilon=first(epsilon), t.usage=first(t.usage), sum.phi=sum(phi)) %>% -->
<!--     mutate(nuc=sapply(template,prevalent_nucleotide), -->
<!--          CG=ifelse(grepl(pattern = 'CG', template), 'y', 'n')) %>% -->
<!--   # ggplot(., aes(sum.phi)) + geom_histogram(bins=40) -->
<!--   mutate(pred.cov=epsilon*(abundance*(sum.phi/(1+sum.phi))))  -->
<!--   # ggplot(., aes(log(t.usage/sum(t.usage)), log(abundance/sum(abundance)))) + geom_point(alpha=0.4, aes(color=nuc, shape=CG)) -->

<!-- cele.pred.cov %>% -->
<!--   ggplot(., aes(log(t.usage/sum(t.usage)), log(pred.cov/sum(pred.cov)))) + geom_point(alpha=0.4, aes(color=nuc, shape=CG)) -->

<!-- cele.pred.cov %>% -->
<!--   ggplot(., aes(log(t.usage), log(abundance))) + geom_point(alpha=0.4, aes(color=nuc, shape=CG)) -->



<!-- zf.pred.cov <- zf.keqs %>% -->
<!--   mutate(abundance=abundance/sum(abundance)) %>% -->
<!--   # mutate(phi=p/-single.keq) %>% -->
<!--   mutate(phi=p/-single.keq) %>% -->
<!--   # ggplot(., aes(phi)) + geom_histogram(bins=40) -->
<!--   group_by(template) %>% -->
<!--   summarise(abundance=first(frac.abundance), epsilon=first(epsilon), t.usage=first(t.usage), sum.phi=sum(phi)) %>% -->
<!--     mutate(nuc=sapply(template,prevalent_nucleotide), -->
<!--          CG=ifelse(grepl(pattern = 'CG', template), 'y', 'n')) %>% -->
<!--   # ggplot(., aes(sum.phi)) + geom_histogram(bins=40) -->
<!--   mutate(pred.cov=epsilon*(abundance*(sum.phi/(1+sum.phi)))) %>% -->
<!--   mutate(pred.cov.frac=pred.cov/sum(pred.cov)) -->
<!--   # ggplot(., aes(log(t.usage/sum(t.usage)), log(abundance/sum(abundance)))) + geom_point(alpha=0.4, aes(color=nuc, shape=CG)) -->
<!-- zf.pred.cov %>% -->
<!--   ggplot(., aes(log(t.usage/sum(t.usage)), log(pred.cov/sum(pred.cov)))) + geom_point(alpha=0.4, aes(color=nuc, shape=CG)) -->

<!-- zf.pred.cov %>% -->
<!--   ggplot(., aes(log(t.usage/sum(t.usage)), log(abundance/sum(abundance)))) + geom_point(alpha=0.4, aes(color=nuc, shape=CG)) -->

<!-- summary(lm(log(t.usage) ~ log(pred.cov.frac), data=zf.pred.cov)) -->
<!-- summary(lm(log(t.usage) ~ log(abundance), data=zf.pred.cov)) -->
<!-- ``` -->
<!-- ```{r} -->
<!-- primer.usage.cele <- cele.all.df %>% -->
<!--   group_by(primer) %>% -->
<!--   summarise(p.usage= sum(pt)) %>% -->
<!--   rename(hex=primer) -->

<!-- cele.df %>% -->
<!--   dplyr::select(template, t.usage,pt, abundance) %>% -->
<!--   rename(hex=template) %>% -->
<!--   inner_join(., primer.usage.cele, by='hex') %>% -->
<!--   mutate(nuc=sapply(hex,prevalent_nucleotide), -->
<!--          CG=ifelse(grepl(pattern = 'CG', hex), 'y', 'n')) %>% -->
<!--   ggplot(., aes(log(t.usage), log(p.usage))) + -->
<!--   geom_point(alpha=0.4, aes(color=nuc, shape=CG)) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- prob <- table(simulate.primer.pool(pool.size=100000, pG=0.05, pT=0.45))/100000 -->
<!-- prob.df <- rename(data.frame(prob), primer=Var1, p=Freq) -->

<!-- prob <- sapply(unique(cele.df$template), primer.prob, probs=c(A=0.25, C=0.25, G=0.05, T=0.45)) -->
<!-- prob.df <- as.data.frame(prob,row.names = F)  -->
<!-- prob.df <- mutate(prob.df, primer=rownames(prob.df)) %>%   -->
<!--   rename(p=prob) -->

<!-- df <- dr.pooled.keqs %>% -->
<!--   dplyr::select(template, primer, single.keq, epsilon) %>% -->
<!--   inner_join(., d3t.all.df, by=c('template', 'primer') ) %>% -->
<!--   inner_join(., prob.df, by=c('primer') ) %>% -->
<!--   mutate(p = 4/p) %>% -->
<!--   mutate(frac.abundance=abundance/tot.ab) %>% -->
<!--   filter(pt>15)  -->
<!--   # mutate(abundance=abundance/sum(abundance)) %>% -->
<!-- df %>% -->
<!--   mutate(phi=p/-single.keq) %>% -->
<!--   # head -->
<!--     group_by(template) %>% -->
<!--   summarise(abundance=first(frac.abundance), epsilon=first(epsilon), t.usage=first(t.usage), sum.phi=sum(phi)) %>% -->
<!--   mutate(pred.cov=epsilon*(abundance*(sum.phi/(1+sum.phi)))) %>% -->
<!--   ggplot(., aes((t.usage/sum(t.usage)), (pred.cov/sum(pred.cov)))) +  -->
<!--   geom_abline(slope=1, intercept=0, color='red') + -->
<!--   geom_point(alpha=0.2)  -->

<!-- ``` -->


<!-- ```{r} -->
<!-- cele.keqs.myway <- compute.keqs.fixedP(cele.filt.df,prob=rename(primer.usage.cele, primer=hex, p=p.usage), mean.eps = filter(species.epsilons, species=='cele')$eps, take.pairs = T) -->
<!-- zf.keqs.myway <- compute.keqs.fixedP(zf.filt.df,prob=rename(primer.usage.zf, primer=hex, p=p.usage), mean.eps = filter(species.epsilons, species=='zf')$eps, take.pairs = T) -->
<!-- human.keqs.myway <- compute.keqs.fixedP(human.filt.df, mean.eps = filter(species.epsilons, species=='human')$eps, take.pairs = T) -->

<!-- zf.keqs.myway %>% -->
<!--   mutate(phi=p/-single.keq) %>% -->
<!--   # ggplot(., aes(phi)) + geom_histogram(bins=40) -->
<!--   group_by(template) %>% -->
<!--   summarise(abundance=first(frac.abundance), epsilon=first(epsilon), t.usage=first(t.usage), sum.phi=sum(phi)) %>% -->
<!--     mutate(nuc=sapply(template,prevalent_nucleotide), -->
<!--          CG=ifelse(grepl(pattern = 'CG', template), 'y', 'n')) %>% -->
<!--   # ggplot(., aes(sum.phi)) + geom_histogram(bins=40) -->
<!--   mutate(pred.cov=epsilon*(abundance*(sum.phi/(1+sum.phi)))) %>% -->
<!--   mutate(pred.cov.frac=pred.cov/sum(pred.cov)) %>% -->
<!--   ggplot(., aes(log(t.usage), log(pred.cov))) + geom_point(alpha=0.3) -->

<!-- ``` -->


<!-- ```{r} -->
<!-- primer.usage.d3r <- d3r.all.df %>% -->
<!--   group_by(primer) %>% -->
<!--   summarise(p= sum(pt))  -->
<!-- d3r.keqs.myway <- compute.keqs.fixedP(d3r.filt,mean.eps = 1000, take.pairs = F)  -->

<!-- d3r.pred.cov <- d3r.keqs.myway %>% -->
<!--   # mutate(p = 4/4^6) %>% -->
<!--   filter(pt>20) %>% -->
<!--   # mutate(abundance=abundance/sum(abundance)) %>% -->
<!--   mutate(phi=p*-single.keq) %>% -->
<!--   # ggplot(., aes(phi)) + geom_histogram(bins=40) -->
<!--     group_by(template) %>% -->
<!--   summarise(abundance=first(frac.abundance), epsilon=first(epsilon), t.usage=first(t.usage), sum.phi=sum(phi)) %>% -->
<!--     # ggplot(., aes(sum.phi)) + geom_histogram(bins=40) -->
<!--   # group_by(template) %>% -->
<!--   # # mutate(pred.cov=epsilon*(abundance*(sum.phi/(1+sum.phi)))) %>% -->
<!--   mutate(pred.cov=epsilon*(abundance*(sum.phi/(1+sum.phi)))) %>% -->
<!--   mutate(nuc=sapply(template,prevalent_nucleotide), -->
<!--          CG=ifelse(grepl(pattern = 'CG', template), 'y', 'n'))  -->

<!-- d3r.pred.cov %>% -->
<!--   # ggplot(., aes(log(t.usage/sum(t.usage)), log(abundance/sum(abundance)))) +  -->
<!--   ggplot(., aes(log(pred.cov/sum(pred.cov)), log(t.usage/sum(t.usage)))) + -->
<!--   theme_classic() + -->
<!--   geom_point(alpha=0.4, aes(color=nuc, shape=CG)) + -->
<!--     # ggplot(., aes(log(t.usage), log(pred.cov))) +  -->
<!--   geom_abline(slope=1, intercept=0, color='red')  -->

<!-- cor(d3r.pred.cov$pred.cov, d3r.pred.cov$t.usage) -->

<!-- cele.all.df %>% -->
<!--   filter(pt>250) %>% -->
<!--   mutate(a=pt/(abundance*1000-t.usage)) %>% -->
<!--   # ggplot(., aes(a)) + geom_histogram(bins=50)  -->
<!--   ggplot(., aes(dG/0.59, -log(a))) + geom_point(alpha=0.4) -->
<!-- ```  -->